src/docs/background/16-off-session-payments/index.md

```diff
@@ -0,0 +1,64 @@
+---
+title: Automatic off-session payments in transaction process
+slug: off-session-payments-in-transaction-process
+updated: 2019-07-30
+category: background
+ingress:
+  With off-session payments you can automatically charge your customers closer
+  to their booking times. This allows for money to be held in Stripe throughout
+  the booking period even when the booking is placed long time in advance.
+published: true
+---
+
+In a typical transaction, the customer pays upfront, but the money is held until
+the transaction completes (e.g. until the booking end date) before it is paid
+out to the provider's bank account. Normally, the maximum amount of time the
+money can be held when using Stripe payments is 90 days. Therefore, the maximum
+amount of time customers can book in advance is limited, if your transaction
+process follows this payment pattern.
+
+Flex API now has capabilities for [saving payment card details for future
+use](https://www.sharetribe.com/api-reference/#stripe-customer). In addition, it
+is now possible to configure your transaction process in such a way, that the
+customer is charged automatically off-session at certain point in time (i.e.
+when they are not present and interacting with your web site or app), provided
+that they have saved a payment card to their account. This allows you to charge
+customers closer to their booking dates, so that the money can be held in Stripe
+throughout the booking period.
+
+## Transaction process example
+
+Suppose your sauna rentals marketplace should allow customers to book saunas up
+to a year in advance, but the customer is charged only a certain time in advance
+before the booking. The figure below illustrates how a part of your transaction
+process might look like.
+
+![Example transaction process with delayed payment](tx-delayed-payments.png 'Example transaction process with delayed payment')
```

Comment:
Comment: In the example process, the "pending-payment" is the state where the client app should prompt the customer to pay, right? But that's also the same state where we try to do the automatic payment transition. So, doesn't this lead to potential race condition? E.g. the user happens to open the app right at the same time when the transition "mark-pending-payment" completes and then he tries to make a payment in the UI, but that transition will fail because automatic "auto-payment" transition is running and locking the transition.

Would it make a bit more robust if the transitions were like this:

- From `accepted` via `auto-payment` to `paid`, 30 days and 10 minutes before booking starts
- From `accepted` via `mark-pending-payment` to `pending-payment`, 30 days before booking starts

In practice it would change the process so that `mark-pending-payment` state is only reached when the automatic payment fails.

Reply:
@rap1ds you are correct, there is a race condition. However, since `mark-pending-payment` is happening when the customer isn't necessarily on the website, I think the risk is small.

What you suggest is unfortunately not possible in Tegel at the moment: what you described is two delayed transitions leaving the same state. You can currently define those like that, but the execution in the engine is that both times are computed and *only the earlier* transition is scheduled. I.e. right now you can't use second delayed transition is a poor mans workaround for "if ... else ...". We discussed implementing this capability briefly with @ovan but we weren't sure allowing multiple delayed transitions is the right solution here in the first place.